File: apps/backend-node/src/shared/storage/storage-url.resolver.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import type { AppConfiguration } from '../../config/configuration';

const PROPERTY_IMAGES_BUCKET = 'property-images';

@Injectable()
export class StorageUrlResolver {
  private readonly supabaseUrl: string;

  constructor(config: ConfigService<AppConfiguration, true>) {
    this.supabaseUrl = config.get('supabaseUrl', { infer: true });
  }

  resolvePropertyImageUrl(storagePath: string | null | undefined): string | null {
    if (!storagePath || storagePath.trim() === '') {
      return null;
    }

    if (
      storagePath.startsWith('http://') ||
      storagePath.startsWith('https://')
    ) {
      return storagePath;
    }

    const base = this.supabaseUrl.replace(/\/+$/, '');
    const path = storagePath.replace(/^\/+/, '');
    return `${base}/storage/v1/object/public/${PROPERTY_IMAGES_BUCKET}/${path}`;
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/shared/StorageUrlResolver.java
package com.housingplatform.shared;

import com.housingplatform.config.AppProperties;
import org.springframework.stereotype.Component;

@Component
public class StorageUrlResolver {

  private static final String PROPERTY_IMAGES_BUCKET = "property-images";

  private final AppProperties appProperties;

  public StorageUrlResolver(AppProperties appProperties) {
    this.appProperties = appProperties;
  }

  public String resolvePropertyImageUrl(String storagePath) {
    if (storagePath == null || storagePath.isBlank()) {
      return null;
    }
    if (storagePath.startsWith("http://") || storagePath.startsWith("https://")) {
      return storagePath;
    }
    String base = appProperties.supabaseUrl().replaceAll("/+$", "");
    String path = storagePath.replaceAll("^/+", "");
    return base + "/storage/v1/object/public/" + PROPERTY_IMAGES_BUCKET + "/" + path;
  }
}
